### PR TITLE
Test for init event

### PR DESCRIPTION
**Pull request by milkovsky**

### CODE REVIEW

**Comment by milkovsky** [comment]
Work is in progress. Not ready to review

File: tests/src/Kernel/EventIntegrationTest.php [hunk 1]
```diff
@@ -157,4 +157,32 @@ public function testSystemLoggerEvent() {
...
+   */
+  public function testInitEvent() {
+    $rule = $this->expressionManager->createRule();
+    $rule->addCondition('rules_test_true');
+    $rule->addAction('rules_test_log');
+
+    $config_entity = $this->storage->create([
+      'id' => 'test_rule',
+      'expression_id' => 'rules_rule',
+      'event' => 'rules_system_logger_event',# @todo which event should we use?
```
Comment (milkovsky) [inline tests/src/Kernel/EventIntegrationTest.php]: which event should we use here?

File: tests/src/Kernel/EventIntegrationTest.php [hunk 2]
```diff
@@ -157,4 +157,32 @@ public function testSystemLoggerEvent() {
...
+    ]);
+    $config_entity->save();
+
+    // Rebuild the container so that the newly configured event gets picked up.
+    $this->container->get('kernel')->rebuildContainer();
+    // The logger instance has changed, refresh it.
+    $this->logger = $this->container->get('logger.channel.rules');
+
+    # @todo trigger Drupal init here.
+    #$this->container->get('cron')->run();
```
Comment (milkovsky) [inline tests/src/Kernel/EventIntegrationTest.php]: How to trigger Drupal initialization here?

**Comment by milkovsky** [comment]
I have some issues with test. I do not understand so far how can we test Drupal initialization when it is already working.
The 'init' event is not a Rules event but kernel.request according to rules.rules.events.yml

File: tests/src/Kernel/EventIntegrationTest.php (hunk 2 above)
Comment (fago) [inline tests/src/Kernel/EventIntegrationTest.php]: no way, just manually trigger the event somehow. e.g. get the dispatcher and dispatch

File: tests/src/Kernel/EventIntegrationTest.php (hunk 1 above)
Comment (fago) [inline tests/src/Kernel/EventIntegrationTest.php]: the init event. look up the name in rules.rules_events.yml file

**Comment by stephenpurkiss** [comment]
Here's some code which may be useful: https://www.drupal.org/node/2013014 has an example of creating a request, and links to https://www.drupal.org/sandbox/sam/2486229 which exposes as hook_init so shows how the symfony kernel.request event relates to the old hook_init rule.

HTH & thanks for picking this up!

**Comment by milkovsky** [comment]
Thank you. Steve! As I understand we should only create a test, right?

**Comment by stephenpurkiss** [comment]
Yes, because kernel.request is already defined as an event because it's just there - as the link above explains there is no hook_init, that is the equivalent of firing off a kernel.request which we can now do at any time as the example on that link shows with the redirect. So I am presuming we need to use code similar to that, or as Fago says dispatch, because that'll set off the event, if it doesn't then the test needs to fail. The gap in between is the code we need, which like you I hadn't figured out just yet but am sure you will, I need to carry on job hunting ;)

File: tests/src/Kernel/EventIntegrationTest.php (hunk 1 above)
Comment (milkovsky) [inline tests/src/Kernel/EventIntegrationTest.php]: The init event does not exist. But there is kernel.request

Comment (stephenpurkiss) [inline tests/src/Kernel/EventIntegrationTest.php]: Please read the links I sent you, they explain that the init event does not exist and the comment on the link goes to a sandbox showing code on how to mimic the init event in 8.

Comment (stephenpurkiss) [inline tests/src/Kernel/EventIntegrationTest.php]: ...I.e. You need to test kernel.request event

**Comment by milkovsky** [comment]
Thanks a lot, good luck :)

**Comment by milkovsky** [comment]
Needs review

File: tests/src/Kernel/EventIntegrationTest.php [hunk 3]
```diff
@@ -157,4 +158,41 @@ public function testSystemLoggerEvent() {
...
+   */
+  public function testInitEvent() {
+    $rule = $this->expressionManager->createRule();
+    $rule->addCondition('rules_test_true');
+    $rule->addAction('rules_test_log');
+
+    $config_entity = $this->storage->create([
+      'id' => 'test_rule',
+      'expression_id' => 'rules_rule',
+      'event' => KernelEvents::REQUEST,
```
Comment (milkovsky) [inline tests/src/Kernel/EventIntegrationTest.php:172]: KernelEvents::REQUEST stands for 'kernel.request'

**Comment by fago** [comment]
merged it.

**Pull request closed without merging**